# Analyzing web-scraped data

The purpose of this project is to analyze the data from the ebay "Kleinanzeigen", a classified ad section of the German ebay website. In the project I will be cleaning, analyzing and manipulathing the data to understand the metrics of car sales on the website.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding='latin_1')

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# First we do some data-cleaning

### The output below tells us that we have 15 'object' data types which 
are likely string data that will need to be converted to numeric data 
type in order to work with it. The remaining columns (5) are int64 which means we can likely work with these columns as they are.

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
autos.isnull()

dateCrawled   name  seller  offerType  price  abtest  vehicleType  \
0            False  False   False      False  False   False        False   
1            False  False   False      False  False   False        False   
2            False  False   False      False  False   False        False   
3            False  False   False      False  False   False        False   
4            False  False   False      False  False   False        False   
...            ...    ...     ...        ...    ...     ...          ...   
49995        False  False   False      False  False   False        False   
49996        False  False   False      False  False   False        False   
49997        False  False   False      False  False   False        False   
49998        False  False   False      False  False   False        False   
49999        False  False   False      False  False   False        False   

       yearOfRegistration  gearbox  powerPS  model  odometer  \
0                   False    False    False  False     False   
1                   False    False    False  False     False   
2                   False    False    False  False     False   
3                   False    False    False  False     False   
4                   False    False    False  False     False   
...                   ...      ...      ...    ...       ...   
49995               False    False    False  False     False   
49996               False    False    False  False     False   
49997               False    False    False  False     False   
49998               False    False    False  False     False   
49999               False    False    False  False     False   

       monthOfRegistration  fuelType  brand  notRepairedDamage  dateCreated  \
0                    False     False  False              False        False   
1                    False     False  False              False        False   
2                    False     False  False              False        False   
3                    False     False  False              False        False   
4                    False     False  False              False        False   
...                    ...       ...    ...                ...          ...   
49995                False     False  False              False        False   
49996                False     False  False              False        False   
49997                False     False  False              False        False   
49998                False     False  False              False        False   
49999                False     False  False              False        False   

       nrOfPictures  postalCode  lastSeen  
0             False       False     False  
1             False       False     False  
2             False       False     False  
3             False       False     False  
4             False       False     False  
...             ...         ...       ...  
49995         False       False     False  
49996         False       False     False  
49997         False       False     False  
49998         False       False     False  
49999         False       False     False  

[50000 rows x 20 columns]

In [7]:
### Show the column names we will be working with

In [8]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

### Make a copy of the dataframe and start doing some cleaning 

In [9]:
autos_copy = autos.copy()

# Rename columns to more logical names

In [10]:
autos_copy.rename(columns={'yearOfRegistration':'registration_year',
                          'monthOfRegistration':'registration_month',
                          'notRepairedDamage':'unrepaired_damage',
                          'dateCreated':'ad_created'},inplace=True)
autos_copy.head() # print the first 5 rows

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuelType  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Now convert the rest of the column names from camelcase to snakecase

I defined a function using the python regular expression module https://docs.python.org/3/library/re.html and then loop through the column names and perform the function on the names.

### I will define a function to use in the loop below for changing case

In [11]:
def camel_to_snake(text):
    '''This function takes in a string and converts it from camelcase
    to snakecase and returns the regular expression from the python
    re module'''
    import re #import the 'regular expression' module
    str1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', text)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', str1).lower()


### As a test of my new function-call it with a camelCase string

In [12]:
camel_to_snake('offerType')

'offer_type'

# Isolate the columns we want to convert to snake case

In [13]:
cased_cols = autos_copy.columns

## Print the unchanged column names out for review

In [14]:
cased_cols

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

### This loop calls the function to convert to snake case

In [15]:
new_case = [] # a list to hold my converted column names
for col in cased_cols:
    snake = camel_to_snake(col)
    new_case.append(snake)

## Below are the desired changes to the column names

In [16]:
print(new_case) 

['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen']


## Now assign the new column names back to the dataframe columns

### First let's create a new dicionary to pass to the pandas.rename function

https://developmentality.wordpress.com/2012/03/30/three-ways-of-creating-dictionaries-in-python/

Let's use the zip and dict functions to do the work:

https://docs.python.org/3/library/functions.html#zip

https://docs.python.org/3/library/stdtypes.html#dict

In [17]:
# Static lists for purpose of illustration
key = autos.columns # the 'key' for my new dictionary
value = new_case # the 'value' for my new dictionary

new_columns_dict = dict(zip(key, value))

### Now we rename the columns in the dataframe to our new column names

In [18]:
# let's plug in the new dictionary I previously created
autos.rename(columns=new_columns_dict, inplace=True)

In [19]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Let's begin some further analysis

In [20]:
autos.describe(include='all').head(4)

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   

       vehicle_type  registration_year  gearbox  power_ps  model   odometer  \
count         44905            50000.0    47320   50000.0  47242      50000   
unique            8                NaN        2       NaN    245         13   
top       limousine                NaN  manuell       NaN   golf  150,000km   
freq          12859                NaN    36993       NaN   4024      32424   

        registration_month fuel_type       brand unrepaired_damage  \
count              50000.0     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   

                 ad_created  nr_of_pictures  postal_code            last_seen  
count                 50000         50000.0      50000.0                50000  
unique                   76             NaN          NaN                39481  
top     2016-04-03 00:00:00             NaN          NaN  2016-04-07 06:17:27  
freq                   1946             NaN          NaN                    8

### From above we can see:
 'price' and 'odometer' columns have NaN values and should be numeric.
 'registration_year' and 'registration_month' have NaN's and other bad values.
 all of the other columns have a lot of non-sensical datatypes that will need to be fixed before we can move on

### Let's remove non-numeric values from the dataset

In [21]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

### Below we locate and replace unwanted characters in column data

In [22]:
autos = autos.drop(["nr_of_pictures", "seller", "offer_type"], axis=1)

In [23]:
autos["odometer"] = (autos["odometer"]
                          .str.replace("km","") #delete the $
                          .str.replace(",","") #delete the ,
                          .astype(int) # now change the type to int
                          )
autos["price"].head()



autos["price"] = (autos["price"]
                          .str.replace("$","") #delete the $
                          .str.replace(",","") #delete the ,
                          .astype(int) # now change the type to int
                          )
autos["price"].head()



0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

### rename the odometer column and verify it worked

In [24]:
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

In [25]:
autos.head(2)

date_crawled                                        name  price  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD   5000   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   8500   

    abtest vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   

   odometer_km  registration_month fuel_type    brand unrepaired_damage  \
0       150000                   3       lpg  peugeot              nein   
1       150000                   6    benzin      bmw              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08

## Analyze the 'price' column

In [26]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
print(autos['price'].value_counts(sort=True, 
                                  ascending=True).head())

#autos["price"].describe().apply(lambda x: format(x, 'f'))



(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
33980    1
2895     1
846      1
8970     1
20790    1
Name: price, dtype: int64


####  Here I use the '.between' method to reassign the dataframe instead of dropping data with .drop method. In this cell I remove the prices that are less than $1.00  and  greater  than $ 500,000.00 as this is an auction site and it is reasonable to assume that these are reasonable bidding ranges.

In [27]:
autos = autos[autos['price'].between(1,500000)]

#### from the printout below we see that the minimum bid price is one dollar and the maximum bid price is three hundred fifty thousand dollars. This seems to confirm my initial assumptions about bid price.

In [28]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        48565.000000  48565.000000  
mean             5.782251  50975.745207  
std              3.685595  25746.968398  
min              0.000000   1067.000000  
25%              3.000000  30657.000000  
50%              6.000000  49716.000000  
75%              9.000000  71665.000000  
max             12.000000  99998.000000

# Now analyze the odometer column

In [29]:
print(autos['odometer_km'].unique().shape)
print(autos['odometer_km'].describe())
print(autos['odometer_km'].value_counts(sort=True, 
                                  ascending=True).head())

#autos["odometer_km"].describe().apply(lambda x: format(x, 'f'))

(13,)
count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
10000    253
20000    762
30000    780
40000    815
5000     836
Name: odometer_km, dtype: int64


#### The information above gives us the min, max and mean values for the odometer_km column and confirms the cleanup actions in the previous cell.

# Analyze the date columns

In [30]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [31]:


(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )



2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [32]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

In [34]:
date_seen = (autos['last_seen'].str[:10])
date_seen.head()

0    2016-04-06
1    2016-04-06
2    2016-04-06
3    2016-03-15
4    2016-04-01
Name: last_seen, dtype: object

In [35]:


(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )



2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64



The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.


In [36]:


print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )



(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [37]:


autos["registration_year"].describe()



count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

# Let's analyze the registration year values

#### Let's get the percentage of years between 1900 and 2016 to determine if we can safely delete those rows. If it's a fairly low percentage we will delete them.

In [38]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

#### Since it's about 4% we will get rid of them and focus on 2016 forward and thus remove the errant year entries.

In [39]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

In [40]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count   46681.000000       46681.000000  46681.000000   46681.000000   
mean     5977.716801        2002.910756    117.892933  125586.855466   
std      9177.909479           7.185103    184.922911   39852.528628   
min         1.000000        1910.000000      0.000000    5000.000000   
25%      1250.000000        1999.000000     75.000000  100000.000000   
50%      3100.000000        2003.000000    109.000000  150000.000000   
75%      7500.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        2016.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        46681.000000  46681.000000  
mean             5.827125  51097.434181  
std              3.670300  25755.387192  
min              0.000000   1067.000000  
25%              3.000000  30827.000000  
50%              6.000000  49828.000000  
75%              9.000000  71732.000000  
max             12.000000  99998.000000

### First identify the brands that have the greatest market share
Here I use value_counts to create index labels of the brands then I will use those labels to create a key value for my aggregation table

In [41]:
brands = autos['brand'].value_counts(normalize=True)
brands_count = brands[brands > .05].index
print(brands_count)


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


### Now loop through the data to perform my aggregation

In [42]:
brand_mean_prices = {}

for brand in brands_count:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)
    

brand_mean_prices 

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

####  Of the top 5 brands, there is a distinct price gap:


    Audi, BMW and Mercedes Benz are more expensive
    Ford and Opel are less expensive
    Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.



### convert to series then dataframe so I can pretty-print the result

In [43]:
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
dtype: int64


In [44]:
df_bmp = pd.DataFrame(bmp_series, columns=['mean_price'])
df_bmp

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

### Now let's calculate the mean mileage for the top brands from before

In [45]:
brand_mean_mileage = {}

for brand in brands_count:
    brand_only = autos[autos["brand"] == brand]
    mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mileage)
    

brand_mean_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

### convert to series then dataframe so I can pretty-print the result

In [46]:
bmm_series = pd.Series(brand_mean_mileage)
print(bmm_series)

volkswagen       128707
bmw              132572
opel             129310
mercedes_benz    130788
audi             129157
ford             124266
dtype: int64


In [47]:
df_bmm = pd.DataFrame(bmm_series, columns=['odometer_km'])
df_bmm

odometer_km
volkswagen          128707
bmw                 132572
opel                129310
mercedes_benz       130788
audi                129157
ford                124266

# Further analysis and conclusion

It appears that while the top five brands fetch the highest price they also have the highest miles of use on them as well. From thia analysis it seems that buyers prefer brand over wear and tear, hence high miles over perceived lesser brands with lower mileage. Is this vanity over practicality or do these brands do as well as th other brands that have lower mileage on them. It would be interesting to investigate maintainance and repair cocst for the top-popular brands versus the less popular brands for less sales price.